materialize-bigquery/binding.go

```diff
@@ -1,225 +1,369 @@
 }
 
-// newBinding generates the bindings for the spec to the BigQuery table.
-func newBinding(generator sqlDriver.Generator, bindingPos int, targetName string, spec *pf.MaterializationSpec_Binding) (*binding, error) {
+// FilePath for the CloudStorage object that the underlying `writer` is connected to.
+func (b *Binding) FilePath() string {
+	obj := b.writer.object
 
-	var err error
-	var b = &binding{
-		name: targetName,
+	return fmt.Sprintf("%s/%s", obj.BucketName(), obj.ObjectName())
+}
+
+func (b *Binding) GenerateDocument(key, values tuple.Tuple, flowDocument json.RawMessage) map[string]interface{} {
+	document := make(map[string]interface{})
+	for i, value := range key {
+		document[b.spec.FieldSelection.Keys[i]] = value
 	}
 
-	// Generate the table definition for this materialization.
-	var tableDef = sqlDriver.TableForMaterialization(targetName, "", generator.IdentifierRenderer, spec)
+	for i, value := range values {
+		document[b.spec.FieldSelection.Values[i]] = value
+	}
 
-	// BINDING LOADS: Load is done via query against an external table with primary keys.
+	if b.spec.FieldSelection.Document != "" {
+		document[b.spec.FieldSelection.Document] = string(flowDocument)
+	}
 
-	// QueryOnPrimaryKey will return the paramsConverter for the primary keys which is all we need here
-	_, b.load.paramsConverter, err = generator.QueryOnPrimaryKey(tableDef)
-	if err != nil {
-		return nil, fmt.Errorf("building load params converter: %w", err)
+	return document
+}
+
+// Job is a small wrapper around `bigquery.Job` and doesn't provide much in terms of logic
+// but ultimately removes some of the boilerplate code for the user's convenience.
+// Since all the data to generate a job lives inside a Binding, it is possible for a Binding
+// to build the job on its own and return a running Job.
+//
+// The returned Job is going to be configured so it's possible to run a query that joins over
+// the bigquery dataset & the cloud storage external table. The name of the table is already computed
+// in the initialization of a Binding and is stored in `externalTableAlias`
+func (b *Binding) Job(ctx context.Context, client *bigquery.Client, q string) (*bigquery.Job, error) {
+	query := client.Query(q)
+	query.TableDefinitions = map[string]bigquery.ExternalData{b.externalTableAlias: &b.writer.edc}
+	query.DefaultDatasetID = b.dataset
+	query.Location = b.region
+
+	return query.Run(ctx)
+}
+
+// Once Commit returns, the binding cannot be written to until Reset() is called.
+func (b *Binding) Commit(ctx context.Context) error {
+	return b.writer.commit(ctx)
+}
+
+// Reset is called when a new Writer need to be instantiated.
+// This is done so every pipeline operation in the bigquery materialization
+// can work with it's own writer to avoid data integrity issues.
+// Call Reset when the lifecycle of a bigquery materialization is completed and
+// the data is committed and acknowledged.
+func (b *Binding) Reset(ctx context.Context, name string) error {
```

Comment:
I'm probably missing parts of the picture here, but is it possible to initialise a new writer during the Store phase? Or put another way, my question is, do we really need to create a writer ahead of our Store phase and keep it in the binding? Can it be a short-lived value inside Store?

Reply:
Not currently, a new writer is only generated at the Prepare phase, or when the transactor is initialized. Similar to the other comment in this PR, this might be refactored if my understanding of "failure mode" is proven wrong.